HRM tests on Node 22 are a bit brittle :-/ The failure is unrelated to my changes and possible race condition. I got another error on my machine just in the first run of tests.

```
 FAIL  playground/hmr/__tests__/hmr.spec.ts > soft invalidate
AssertionError: expected 'soft-invalidation/index.js is transfo…' to match 'soft-invalidation/index.js is transfo…'

Expected: "soft-invalidation/index.js is transformed 2 times. child is now updated?"
Received: "soft-invalidation/index.js is transformed 3 times. child is now updated?"
```

--- a/packages/vite/src/node/__tests__/utils.spec.ts
+++ b/packages/vite/src/node/__tests__/utils.spec.ts
@@ -407,6 +407,21 @@ describe('processSrcSetSync', () => {
     const result = 'url("picture.png") 1x type("image/jpeg")'
     expect(processSrcSetSync(source, ({ url }) => url)).toBe(result)
   })
+
+  test('should capture whole image set options', async () => {
+    const source = `linear-gradient(cornflowerblue, white) 1x,
+                    url("detailed-gradient.png") 3x`
+    const expected = [
+      'linear-gradient(cornflowerblue, white)',
+      'url("detailed-gradient.png")',
+    ]
+    const result: string[] = []
+    processSrcSetSync(source, ({ url }) => {
+      result.push(url)
+      return url
+    })
+    expect(result).toEqual(expected)
+  })
 })
 
 describe('flattenId', () => {

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -727,42 +727,29 @@ function joinSrcset(ret: ImageCandidate[]) {
     .join(', ')
 }
 
-/*!
- * Based on https://github.com/sindresorhus/srcset
- * MIT License, Copyright (c) Sindre Sorhus <sindresorhus@gmail.com> (https://sindresorhus.com)
- */
-/**
- This regex represents a loose rule of an “image candidate string” and "image set options".
-
- @see https://html.spec.whatwg.org/multipage/images.html#srcset-attribute
- @see https://drafts.csswg.org/css-images-4/#image-set-notation
-
-  An “image candidate string” roughly consists of the following:
-  1. Zero or more whitespace characters.
-  2. A non-empty URL that does not start or end with `,`.
-  3. Zero or more whitespace characters.
-  4. An optional “descriptor” that starts with a whitespace character.
-  5. Zero or more whitespace characters.
-  6. Each image candidate string is separated by a `,`.
- */
-// eslint-disable-next-line regexp/no-super-linear-backtracking
-const imageCandidateRegex = /\s*([^,]\S*[^,](?:\s[^,]+)?)\s*(?:,|$)/
 const escapedSpaceCharacters = /(?: |\\t|\\n|\\f|\\r)+/g
 
 export function parseSrcset(string: string): ImageCandidate[] {
-  return string
+  const matches = string
+    .trim()
     .replace(escapedSpaceCharacters, ' ')
     .replace(/\r?\n/, '')
     .replace(/,\s+/, ', ')
-    .split(imageCandidateRegex)
-    .filter((_part, index) => index % 2 === 1)
-    .map((part) => {
-      const [url, ...descriptors] = part.trim().split(/\s+/)
-      // in `image-set()` context descriptor can have `resolution` and `type` parts separated by whitespace,
-      // we need to join them back together
-      return { url, descriptor: descriptors.join(' ') }
-    })
-    .filter(({ url }) => !!url)
+    .replaceAll(/\s+/g, ' ')
+    .matchAll(
+      /**
+       This regex represents a loose rule of an “image candidate string” and "image set options".
+
+       @see https://html.spec.whatwg.org/multipage/images.html#srcset-attribute
+       @see https://drafts.csswg.org/css-images-4/#image-set-notation
+       */
+      // eslint-disable-next-line regexp/no-super-linear-backtracking
+      /\s*(?<url>[\w-]+\([^)]*\)|"[^"]*"|'[^']*'|[^,]\S*[^,])\s*(?<descriptor>[^,]*(?:\stype\([^)]*\))?)\s*(?:,|$)/g,
+    )
+  return Array.from(matches, ({ groups }) => ({
+    url: groups.url.trim(),
+    descriptor: groups.descriptor.trim(),
+  })).filter(({ url }) => !!url)
 }
 
 export function processSrcSet(

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -727,6 +727,22 @@ function joinSrcset(ret: ImageCandidate[]) {
     .join(', ')
 }
 
+/**
+ This regex represents a loose rule of an “image candidate string” and "image set options".
+
+ @see https://html.spec.whatwg.org/multipage/images.html#srcset-attribute
+ @see https://drafts.csswg.org/css-images-4/#image-set-notation
+
+  The Regex has named capturing groups `url` and `descriptor`.
+  The `url` group can be:
+  * any CSS function
+  * CSS string (single or double-quoted)
+  * or URL string.
+  The `descriptor` is anything after the space and before the comma,
+  and can have optional `type(...)` for the `image-set` case.
+ */
+const imageCandidateRegex =
+  /(?:^|\s)(?<url>[\w-]+\([^)]*\)|"[^"]*"|'[^']*'|[^,]\S*[^,])\s*(?:\s(?<descriptor>[^,\s]+(?:\stype\([^)]*\))?)\s*)?(?:,|$)/g
 const escapedSpaceCharacters = /(?: |\\t|\\n|\\f|\\r)+/g
 
 export function parseSrcset(string: string): ImageCandidate[] {
@@ -736,16 +752,7 @@ export function parseSrcset(string: string): ImageCandidate[] {
     .replace(/\r?\n/, '')
     .replace(/,\s+/, ', ')
     .replaceAll(/\s+/g, ' ')
-    .matchAll(
-      /**
-       This regex represents a loose rule of an “image candidate string” and "image set options".
-
-       @see https://html.spec.whatwg.org/multipage/images.html#srcset-attribute
-       @see https://drafts.csswg.org/css-images-4/#image-set-notation
-       */
-      // eslint-disable-next-line regexp/no-super-linear-backtracking
-      /\s*(?<url>[\w-]+\([^)]*\)|"[^"]*"|'[^']*'|[^,]\S*[^,])\s*(?<descriptor>[^,]*(?:\stype\([^)]*\))?)\s*(?:,|$)/g,
-    )
+    .matchAll(imageCandidateRegex)
   return Array.from(matches, ({ groups }) => ({
     url: groups?.url?.trim() ?? '',
     descriptor: groups?.descriptor?.trim() ?? '',

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -747,8 +747,8 @@ export function parseSrcset(string: string): ImageCandidate[] {
       /\s*(?<url>[\w-]+\([^)]*\)|"[^"]*"|'[^']*'|[^,]\S*[^,])\s*(?<descriptor>[^,]*(?:\stype\([^)]*\))?)\s*(?:,|$)/g,
     )
   return Array.from(matches, ({ groups }) => ({
-    url: groups.url.trim(),
-    descriptor: groups.descriptor.trim(),
+    url: groups?.url?.trim() ?? '',
+    descriptor: groups?.descriptor?.trim() ?? '',
   })).filter(({ url }) => !!url)
 }
 

